src/relax/op/op_common.h

```diff
@@ -33,8 +33,172 @@
+    output_dtype = DataType::Void();
+  } else if (lhs_sinfo->dtype != rhs_sinfo->dtype) {
+    ctx->ReportFatal(Diagnostic::Error(call)
+                     << "Data types " << lhs_sinfo->dtype << " and " << rhs_sinfo->dtype
+                     << " must be equal for broadcasting operators");
+  } else {
+    output_dtype = lhs_sinfo->dtype;
+  }
+  return output_dtype;
+};
+
+/*!
+ * \brief Check if the given layout exactly contains all the specified characters.
+ * \param layout The input layout to be checked
+ * \param dims The given characters that are required to contain
+ * \return A boolean indicating if the input layout exactly contains all the specified characters.
+ */
+inline bool CheckTensorLayout(const String& layout, std::vector<char> dims) {
+  if (layout.size() != dims.size()) {
+    return false;
+  }
+  std::string _layout = layout.operator std::string();
+  for (char dim : dims) {
+    if (_layout.find(dim) == std::string::npos) {
+      return false;
+    }
+  }
+  return true;
+}
+
+/*!
+ * \brief Complete the padding to a 4-length array.
+ * - If the padding length is 1, the same padding is used on all top/left/bottom/right sides
+ * - If the padding length is 2, top/bottom sides use padding[0] and left/right use padding[1]
+ * - If the padding length is 4, padding is in the order of (top, left, bottom, right)
+ * \param padding The given padding to be completed
+ * \return The completed padding.
+ * \throws Throws error if the input padding length is neither 1, 2 or 4.
+ */
+inline Array<PrimExpr> GetCompletePadding2D(Array<PrimExpr> padding) {
+  if (padding.size() == 1) {
+    return {padding[0], padding[0], padding[0], padding[0]};
+  } else if (padding.size() == 2) {
+    return {padding[0], padding[1], padding[0], padding[1]};
+  } else if (padding.size() == 4) {
+    return padding;
+  }
+  LOG(FATAL) << "The input padding length is expected to be either 1, 2 or 4. However, the given "
+                "padding is "
+             << padding;
+  throw;
+}
+
+/*!
+ * \brief Check if the given tensor struct info has the given expected ndim (or the ndim is
+ * unknown), and try to cast the shape to ShapeExpr.
+ * \param call The Call of the binary tensor operators.
+ * \param ctx The error reporting context.
+ * \param sinfo The input tensor struct info to be checked.
+ * \param expected_ndim The ndim that the input tensor struct info is expected to have
+ * \param op_name The name of the operator that invokes this function.
+ * \return The shape of the input tensor in ShapeExpr, or `NullOpt` if the shape is unknown.
+ */
+inline Optional<ShapeExpr> CheckNdimAndGetShape(const Call& call, const BlockBuilder& ctx,
+                                                const TensorStructInfo& sinfo, int expected_ndim,
+                                                const String& op_name) {
+  if (!sinfo->IsUnknownNdim() && sinfo->ndim != expected_ndim) {
+    ctx->ReportFatal(Diagnostic::Error(call->span)
```

Comment:
in error message, avoid passing in call->span, instead directly pass in call. This would helps us to handle cases where call do not have a span